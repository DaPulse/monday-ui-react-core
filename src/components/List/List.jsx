import React, { useRef, forwardRef, useMemo } from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import useMergeRefs from "../../hooks/useMergeRefs";
import "./List.scss";

const List = forwardRef(({ className, id, component, children, dense, ariaLabel, ariaDescribedBy }, ref) => {
  const componentRef = useRef(null);
  const mergedRef = useMergeRefs({ refs: [ref, componentRef] });
  const Component = useMemo(() => component, [component]);
  return (
    <Component
      ref={mergedRef}
      className={cx("monday-style-list", className, { "monday-style-list--dense": dense })}
      id={id}
      aria-label={ariaLabel}
      aria-describedby={ariaDescribedBy}
    >
      {children}
    </Component>
  );
});

List.propTypes = {
  /**
   * class name to be add to the wrapper
   */
  className: PropTypes.string,
  /**
   * id to be add to the wrapper
   */
  id: PropTypes.string,
  /**
   * the wrapping component to wrap the List Items [div, nav, ul, ol]
   */
  component: PropTypes.string,
  /**
   * remove the side padding
   */
  dense: PropTypes.bool,
  /**
   * ARIA label string to describe to list
   */
  ariaLabel: PropTypes.string,
  /**
   * ARIA described by string to reference an id to describe by
   */
  ariaDescribedBy: PropTypes.string
};
List.defaultProps = {
  className: "",
  id: undefined,
  component: "ul",
  dense: false,
  ariaLabel: undefined,
  ariaDescribedBy: undefined
};

export default List;
