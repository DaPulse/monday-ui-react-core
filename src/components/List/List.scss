@import "../../styles/themes.scss";
@import "../../styles/states.scss";
@import "../../styles/typography.scss";

.monday-style-list {
  margin: 0;
  padding: 0;
  overflow: auto;
  @include scroller();
}