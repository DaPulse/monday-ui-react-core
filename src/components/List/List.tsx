import cx from "classnames";
import React, {
  CSSProperties,
  FC,
  forwardRef,
  ReactElement,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import useMergeRefs from "../../hooks/useMergeRefs";
import useKeyEvent from "../../hooks/useKeyEvent";
import { VirtualizedListItems } from "./VirtualizedListItems/VirtualizedListItems";
import { keyCodes, UP_DOWN_ARROWS } from "../../constants/keyCodes";
import VibeComponentProps from "../../types/VibeComponentProps";
import { ListItemProps } from "../ListItem/ListItem";
import { ListTitleProps } from "../ListTitle/ListTitle";
import { ListWrapperComponentType } from "./ListConstants";
import { ComponentDefaultTestId, getTestId } from "../../tests/test-ids-utils";
import styles from "./List.module.scss";

export interface ListProps extends VibeComponentProps {
  /**
   * the wrapping component to wrap the List Items [div, nav, ul, ol]
   */
  component?: ListWrapperComponentType;
  /**
   * ARIA label string to describe to list
   */
  ariaLabel?: string;
  /**
   * ARIA described by string to reference an id to describe by
   */
  ariaDescribedBy?: string;
  children?: ReactElement<ListItemProps | ListTitleProps> | ReactElement<ListItemProps | ListTitleProps>[];
  /**
   * Using virtualized list for rendering only the items which visible to the user in any given user (performance optimization)
   */
  renderOnlyVisibleItems?: boolean;
  style?: CSSProperties;
}

const List: FC<ListProps> = forwardRef(
  (
    {
      className,
      id,
      component = "ul",
      children,
      ariaLabel,
      ariaDescribedBy,
      renderOnlyVisibleItems = false,
      style,
      "data-testid": dataTestId
    },
    ref
  ) => {
    const componentRef = useRef(null);
    const [focusIndex, setFocusIndex] = useState(0);
    const mergedRef = useMergeRefs({ refs: [ref, componentRef] });
    const Component = component;
    const childrenRefs = useRef([]);

    useEffect(() => {
      if (!id) {
        console.warn("List should have a valid id prop");
      }
    }, [id]);

    const getListItemIdByIndex = (index: number) => {
      return childrenRefs.current[index]?.id;
    };

    const updateFocusedItem = useCallback((index: number) => {
      setFocusIndex(index);
      componentRef?.current?.setAttribute("aria-activedescendant", getListItemIdByIndex(index));
    }, []);

    const onUpDownArrows = useCallback(
      (event: KeyboardEvent) => {
        if (renderOnlyVisibleItems) {
          return;
        }
        event.preventDefault();

        const isUpKey = event.key === keyCodes.UP_ARROW;
        const isDownKey = event.key === keyCodes.DOWN_ARROW;
        let overrideFocusIndex = undefined;

        if (isDownKey && focusIndex + 1 < childrenRefs.current.length) {
          overrideFocusIndex = focusIndex + 1;
        } else if (isUpKey && focusIndex > 0) {
          overrideFocusIndex = focusIndex - 1;
        }
        if (overrideFocusIndex !== undefined) {
          updateFocusedItem(overrideFocusIndex);
          childrenRefs.current[overrideFocusIndex].focus();
        }
      },
      [focusIndex, renderOnlyVisibleItems, updateFocusedItem]
    );

    useKeyEvent({
      keys: UP_DOWN_ARROWS,
      callback: onUpDownArrows,
      ref: componentRef
    });

    const overrideChildren = useMemo(() => {
      let override: ReactElement | ReactElement[] = Array.isArray(children) ? children : [children];
      if (renderOnlyVisibleItems) {
        override = <VirtualizedListItems>{override}</VirtualizedListItems>;
      } else {
        childrenRefs.current = childrenRefs.current.slice(0, override.length);
        override = React.Children.map(override, (child, index) => {
          if (!React.isValidElement(child)) {
            return child;
          }

          return React.cloneElement(child, {
            // @ts-ignore not sure how to deal with ref here
            ref: ref => (childrenRefs.current[index] = ref),
            tabIndex: focusIndex === index ? 0 : -1,
            updateFocusedItem,
            index,
            listId: id
          });
        });
      }

      return override;
    }, [children, focusIndex, id, updateFocusedItem, renderOnlyVisibleItems]);

    return (
      // @ts-ignore Component comes from string, so it couldn't have types
      <Component
        data-testid={dataTestId || getTestId(ComponentDefaultTestId.LIST, id)}
        ref={mergedRef}
        style={style}
        className={cx(styles.list, className)}
        id={id}
        aria-label={ariaLabel}
        aria-describedby={ariaDescribedBy}
        tabIndex={-1}
        role="listbox"
      >
        {overrideChildren}
      </Component>
    );
  }
);

export default List;
