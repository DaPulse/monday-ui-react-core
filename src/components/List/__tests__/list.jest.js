import React from "react";
import renderer from "react-test-renderer";
import List from "../List";

it("renders correctly with empty props", () => {
  const tree = renderer.create(<List>Something</List>).toJSON();
  expect(tree).toMatchSnapshot();
});
