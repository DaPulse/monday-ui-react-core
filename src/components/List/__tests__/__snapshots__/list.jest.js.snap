// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with empty props 1`] = `
<ul
  className="list monday-style-list"
  data-testid="virtualized-list"
  onKeyDown={[Function]}
  tabIndex={-1}
>
  Something
</ul>
`;

exports[`renders correctly without props 1`] = `
<ul
  className="list monday-style-list"
  data-testid="virtualized-list"
  onKeyDown={[Function]}
  tabIndex={-1}
/>
`;
